### DISCUSSION TITLE
It would be nice if TaskTeardown reported success or failure

### DISCUSSION THREAD

**Discussion by grdylan** [cake-build/cake#3732] [Ideas]
Hey guys

I was just working with `TaskTeardown` on a TFS build system to control how TFS reports task progress. The way it works in TFS is that you spit out console commands to tell TFS when tasks complete. For example:

```csharp
TaskTeardown(context => Information($"##vso[task.complete result=Succeeded;]{context.Task.Name}"));
```

Unfortunately, there's not enough information in the `context` to be able to tell if the task was successful or not so we just have to hard code "Succeeded" into the string.

There's a similar issue #2130 that also mentions this. In that issue he was pointed to a workaround on gitter..

> Well if you come to task teardown it's successful, if not it's failure, so you could track in TaskSetup/TaskTeardown

That doesn't work and nor should it in my opinion. It wouldn't really make any sense to *not* call `TaskTeardown` if the task failed. The whole point of a Teardown method is to be able to report or otherwise clean up from a task even if it fails.

This functionality was discussed in the original issue #389 but it wasn't implemented as far as I can tell because:

> it's probably wise to include as little information as possible and let this grow organically.

Fair enough. Maybe it's time to reconsider what information is provided and grow the API a little more.

I believe this functionality is quite useful for controlling build systems. It's quite common to use these kinds of console output commands to do this. TeamCity also works this way.

For reference [here's the horribly formatted TFS documentation on supported commands](https://github.com/Microsoft/vsts-tasks/blob/master/docs/authoring/commands.md). In TFS the valid values for result are:

```
Succeeded
SucceededWithIssues
Failed
Cancelled
Skipped
```

**Comment by pascalberger** [comment]
See [Cake.BuildSystems.Module](https://cakebuild.net/extensions/cake-buildsystems-module/) which provides this kind of integration with different build systems.

Cake 2.0 has also been enhanced with additional events on `CakeEngine` class. See https://github.com/cake-build/cake/issues/3151 for details.
